File: graph_theory/edmond_karp/edmond_karp.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <iterator>
#include "utils/graph.h"

#define RESIDUE(s,t) (capacity[s][t] - flow[s][t])

Weight EdmondKarp(const Graph &g, int s, int t)
{
    int n = g.size();
    Matrix flow(n , Array(n)), capacity(n, Array(n));
    REP(u, n) FOR(e, g[u]) capacity[e->src][e->dst] += e->weight;

    Weight total = 0;
    while (1) {
        std::queue<int> Q; Q.push(s);
        std::vector<int> prev(n, -1); prev[s] = s;
        while (!Q.empty() && prev[t] < 0) {
            int u = Q.front(); Q.pop();
            FOR(e, g[u]) if (prev[e->dst] < 0 && RESIDUE(u, e->dst)) {
                prev[e->dst] = u;
                Q.push(e->dst);
            }
        }
        if (prev[t] < 0) return total;
        Weight inc = INF;
        for (int j = t; prev[j] != j; j = prev[j])
            inc = std::min(inc, RESIDUE(prev[j], j));
        for (int j = t; prev[j] != j; j = prev[j])
            flow[prev[j]][j] += inc, flow[j][prev[j]] -= inc;
        total += inc;
    }
}

int main()
{
    int N, M;
    int src, dst;

    while (std::cin >> N && N) {
        Graph g;
        g.resize(N);

        std::cin >> src >> dst >> M;
        REP(i, M) {
            int x, y, c;
            std::cin >> x >> y >> c;
            g[x].push_back(Edge(x, y, c));
        }

        std::cout << EdmondKarp(g, src, dst) << std::endl;
    }
    return 0;
}

File: graph_theory/edmond_karp/edmond_karp.py
#!/usr/bin/env python
import numpy
import queue

class Edge():
    def __init__(self, src, dst, weight):
        self.weight = weight
        self.src = src
        self.dst = dst


class Graph():
    def __init__(self, edges):
        self.edges = edges



def edmond_karp(g, s, t):
    n = len(g)

    flow = numpy.zeros((n, n))
    capacity = numpy.zeros((n, n))

    for e in g:
        capacity[e.src, e.dst] += e.weight

    total = 0
    while True:
        Q = queue.Queue(maxsize=n)
        prev = [-1 * n]
        prev[s] = s

        while not Q.empty() and prev[t] < 0:
            u = Q.get()
            for e in g[u]:
                if prev[e.dst] < 0 and (capacity[e.src, e.dst] - flow[e.src, e.dst]) > 0:
                    prev[e.dst] = e.src
                    Q.put(e.dst)

        if prev[t] < 0:
            return total

        inc = 9999999
        j = t
        while prev[j] != j:
            if inc > capacity[prev[j], j] - flow[prev[j], j]:
                inc = capacity[prev[j], j] - flow[prev[j], j]

            j = prev[j]


        j = t
        while prev[j] != j:
            flow[prev[j], j] += inc
            flow[j, prev[j]] -= inc

        total += inc



if __name__ == "__main__":
    pass

